src/lm.jl

```diff
@@ -126,39 +143,56 @@ Fit a linear model to data.
 
 $FIT_LM_DOC
 """
+
+function convert_weights(wts)
+    if wts isa Union{FrequencyWeights, AnalyticWeights, ProbabilityWeights, UnitWeights}
+        wts
+    elseif wts isa AbstractVector
+        Base.depwarn("Passing weights as vector is deprecated in favor of explicitly using " *
+                     "`AnalyticWeights`, `ProbabilityWeights`, or `FrequencyWeights`. Proceeding " *
+                     "by coercing `wts` to `FrequencyWeights`", :fit)
+        fweights(wts)
+    else
+        throw(ArgumentError("`wts` should be an `AbstractVector` coercible to `AbstractWeights`"))
+    end
+end
+
 function fit(::Type{LinearModel}, X::AbstractMatrix{<:Real}, y::AbstractVector{<:Real},
              allowrankdeficient_dep::Union{Bool,Nothing}=nothing;
-             wts::AbstractVector{<:Real}=similar(y, 0),
+             wts::Union{AbstractWeights{<:Real}, AbstractVector{<:Real}}=uweights(length(y)),
              dropcollinear::Bool=true,
              method::Symbol=:cholesky)
     if allowrankdeficient_dep !== nothing
         @warn "Positional argument `allowrankdeficient` is deprecated, use keyword " *
               "argument `dropcollinear` instead. Proceeding with positional argument value: $allowrankdeficient_dep"
         dropcollinear = allowrankdeficient_dep
     end
+    # For backward compatibility accept wts as AbstractArray and coerce them to FrequencyWeights
+    _wts = convert_weights(wts)
 
     if method === :cholesky
-        fit!(LinearModel(LmResp(y, wts), cholpred(X, dropcollinear), nothing))
+        fit!(LinearModel(LmResp(y, _wts), cholpred(X, dropcollinear, _wts), nothing))
     elseif method === :qr
-        fit!(LinearModel(LmResp(y, wts), qrpred(X, dropcollinear), nothing))
+        fit!(LinearModel(LmResp(y, _wts), qrpred(X, dropcollinear, _wts), nothing))
     else
         throw(ArgumentError("The only supported values for keyword argument `method` are `:cholesky` and `:qr`."))
     end
 end
 
 function fit(::Type{LinearModel}, f::FormulaTerm, data,
              allowrankdeficient_dep::Union{Bool,Nothing}=nothing;
-             wts::Union{AbstractVector{<:Real}, Nothing}=nothing,
+             wts::Union{AbstractWeights{<:Real}, AbstractVector{<:Real}}=uweights(0),
              dropcollinear::Bool=true,
              method::Symbol=:cholesky,
              contrasts::AbstractDict{Symbol}=Dict{Symbol,Any}())
-    f, (y, X) = modelframe(f, data, contrasts, LinearModel)
-    wts === nothing && (wts = similar(y, 0))
 
+    f, (y, X) = modelframe(f, data, contrasts, LinearModel)
+    _wts = convert_weights(wts)
+    _wts = isempty(_wts) ? uweights(length(y)) : _wts
```

Comment:
Also print a deprecation warning when weights have a different length from `y`. We don't want to continue accepting empty vectors in the future as people should use `UnitWeights` instead.